dissect/cstruct/types/structure.py

```diff
@@ -235,142 +321,360 @@ def _write(self, stream: BinaryIO, data: Instance) -> int:
-                field.alignment = field.type.alignment
+        for field in fields:
+            if size is not None:
+                try:
+                    size = max(len(field.type), size)
+                except TypeError:
+                    size = None
 
-            size = max(len(field.type), size)
             alignment = max(field.alignment, alignment)
 
-        if self.align and size is not None:
+        if align and size is not None:
             # Add "tail padding" if we need to align
             # This bit magic rounds up to the next alignment boundary
             # E.g. offset = 3; alignment = 8; -offset & (alignment - 1) = 5
             size += -size & (alignment - 1)
 
-        self.size = size
-        self.alignment = alignment
+        return size, alignment
 
-    def _read(self, stream: BinaryIO, context: dict[str, Any] = None) -> Instance:
-        buf = io.BytesIO(memoryview(stream.read(len(self))))
-        result = OrderedDict()
+    def _read_fields(cls, stream: BinaryIO, context: dict[str, Any] = None) -> tuple[dict[str, Any], dict[str, int]]:
+        result = {}
         sizes = {}
 
-        for field in self.fields:
-            start = 0
-            buf.seek(0)
-            field_type = self.cstruct.resolve(field.type)
+        if cls.size is None:
+            offset = stream.tell()
+            buf = stream
+        else:
+            offset = 0
+            buf = io.BytesIO(stream.read(cls.size))
+
+        for field in cls.__fields__:
+            field_type = cls.cs.resolve(field.type)
 
-            if field.offset:
-                buf.seek(field.offset)
+            start = 0
+            if field.offset is not None:
                 start = field.offset
 
-            v = field_type._read(buf, result)
+            buf.seek(offset + start)
+            value = field_type._read(buf, result)
 
-            if isinstance(field_type, Structure) and field_type.anonymous:
-                sizes.update(v._sizes)
-                result.update(v._values)
-            else:
-                sizes[field.name] = buf.tell() - start
-                result[field.name] = v
+            sizes[field.name] = buf.tell() - start
+            result[field.name] = value
 
-        return Instance(self, result, sizes)
+        return result, sizes
 
-    def _write(self, stream: BinaryIO, data: Instance) -> Instance:
+    def _read(cls, stream: BinaryIO, context: dict[str, Any] = None) -> Union:
+        if cls.size is None:
+            start = stream.tell()
+            result, sizes = cls._read_fields(stream, context)
+            size = stream.tell() - start
+            stream.seek(start)
+            buf = stream.read(size)
+        else:
+            result = {}
+            sizes = {}
+            buf = stream.read(cls.size)
+
+        obj: Union = cls(**result)
+        object.__setattr__(obj, "_values", result)
+        object.__setattr__(obj, "_sizes", sizes)
+        object.__setattr__(obj, "_buf", buf)
```

Comment:
```suggestion
        setattrs(object, obj, {"_values":result, "_sizes":sizes, "_buf":buf})
```

I see this pattern several times, maybe add a little helper function?

Reply:
I understand why, but I don't think I will. This code is a pretty hot path, and this is more explicit on the internals.